src/Statement/StatementList.php

```diff
@@ -45,10 +45,20 @@ public function __construct( $statements = array() /*...*/ ) {
 			$statements = func_get_args();
 		}
 
+		$this->setStatements( $statements );
+	}
+
+	/**
+	 * @param Statement[]|Traversable
+	 *
+	 * @throws InvalidArgumentException
+	 */
+	public function setStatements( $statements ) {
 		if ( !is_array( $statements ) && !( $statements instanceof Traversable ) ) {
 			throw new InvalidArgumentException( '$statements must be an array or an instance of Traversable' );
 		}
 
+		$this->statements = array();
```

Comment:
Seems like you'll either need more memory or do the loop twice. Not sure which very minor point is more important :)